scalyr_agent/monitors_manager.py

```diff
@@ -151,6 +169,46 @@ def stop_manager(self):
             except:
                 log.exception('Failed to close the metric log due to an exception')
 
+        self.stop(wait_on_join=wait_on_join, join_timeout=join_timeout)
+
+    def set_user_agent_augment_callback(self, callback):
+        """Registers a callback that, when invoked, will augment the User-Agent.
+
+        The callback should be passed a single list of string fragments which will be appended in order.
+        """
+        self.__user_agent_callback = callback
+
+    def run(self):
+        """Poll monitors for specific state changes (e.g. User-Agent fragment)"""
+
+        prev_user_agent_frags = None
+
+        # noinspection PyBroadException
+        try:
+            while self._run_state.is_running():
+                self.__lock.acquire()
+                # noinspection PyBroadException
+                try:
+                    # Invoke user-agent callback only if the fragments have changed
+                    user_agent_frags = set()
+                    for monitor in self.__running_monitors:
+                        frag = monitor.get_user_agent_fragment()
+                        if frag:
+                            user_agent_frags.add(frag)
+                    if user_agent_frags != prev_user_agent_frags:
+                        self.__user_agent_callback(sorted(user_agent_frags))
+                    prev_user_agent_frags = user_agent_frags
+                except Exception:
+                    log.exception('Monitor manager failed to query monitor %s' % monitor.monitor_name)
+                finally:
+                    self.__lock.release()
+                self._run_state.sleep_but_awaken_if_stopped(MonitorsManager.POLL_MONITORS_SECS)
```

Comment:
BTW, should we make this a generic interval where monitors are polled for all sorts of information (right now it's only user-agent, but in the future it could be be more). Or do we want the different settings for each type of information pollable from monitors? (If so, we would have to keep track of elapsed time in the manager poll loop and then selectively poll different types of info as appropriate to the individual settings.

Reply:
My suggestion is that we don't over-generalize too early (kind of like not optimizing too early).  Let's just code up this specific feature in the way that makes sense for it to be coded up.  So, let's use `user_agent_refresh_interval` since it communicates to the user that it does, etc.  After all, this is the first time in 4 years that we are adding something like this, so it's not clear to me if we will need a generic mechanism for these type of things.